### DISCUSSION TITLE
examples for custom html validation error messages?

### DISCUSSION THREAD

**Discussion by eleith** [ariakit/ariakit#1783] [Questions]
are there existing patterns to be able to use the `Form` component while also [customizing the native error messages ](https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation#validating_forms_using_javascript) that are triggered when using HTML attributes like `required`, `pattern` and others?

i've tried a few solutions and they are pretty cumbersome. it is also considerably more cumbersome when trying to customize these messages with `FormField` + `Select`.

currently, the cleanest workaround is to avoid using validation attributes and implement them all with `useValidate`. 

some of the ergonomics might be improved if the form `state` made it more convenient to access the field refs and their `validity: ValidityState` object.

**Comment by diegohaz** [comment]
The native validation is meant to be used for simple use cases. That's a relatively cheap feature to support right now. I'm not sure we should add more features on top of that. Internally, we use `useValidate`:

https://github.com/ariakit/ariakit/blob/182ff27f030ebe12994690e72387d15f486baf3e/packages/ariakit/src/form/form-field.ts#L104-L112

That's pretty much all the code we need to support built-in validation.

If you need custom messages, I'd suggest using `useValidate` with a custom validation logic.

**Reply by diegohaz** [reply]
We could add something like this:

```jsx
<FormField
  getValidationMessage={(validity: ValidityState) => {
    return "custom message";
  }}
/>
```

I don't know if that would be that common, though. But feel free to open a feature request issue if you think it's worth it, so others can chime in and give their opinions.

**Comment by eleith** [comment]
@diegohaz thanks for the suggestions.

my early intuition was just to set my own error message, but since the default one is being set internally, there is some race condition of both messages flipping between each other. (click in and out of the name field in https://codesandbox.io/s/bold-breeze-rkdthn?file=/src/form.tsx) 

this sort of makes sense once i read the code, but it wasn't expected when i was working only at the API level.

instead, one needs to override the messaging using the native `setCustomValidity` (and then unset it when the condition changes, as seen in the title field in https://codesandbox.io/s/bold-breeze-rkdthn?file=/src/form.tsx)

i like the ergonomics of `form.setError` and `form.setErrors` but if one were to use `required`, `pattern` or other native constraints, then you either have to accept the default messages, not use native constraints or do as i did in the sandbox and get at the underlying dom element to set and unset the custom messages.

i think if `form` made it easier to access the underlying refs that might go along way. you can do this by looping through `form.items` but that includes inputs and labels. a `form.refs.name` and `form.refs.title` would make it slightly easier. 

though, in the the end, i think the lack of intuition that `setError` would conflict instead of override the native error message is the culprit of this discussion.
